test/partition-healing-tests.js

```diff
@@ -0,0 +1,377 @@
+                //   sut's incarnation number to 1000.
+                // - According to sut, A's incarnation number needs to be higher than according to B.
+                //   We do this by splitting A's and B's membership information.
+
+                // Mark b1 and b2 faulty in a1. Strictly for this test, there is no need to update
+                // a1's membership, but we make it for consistency/understandability.
+                markBFaultyInA(t, tc),
+
+                // Mark sut and a1 faulty in memberships of B.
+                markSutA1FaultyInB(t, tc),
+
+                // 2. Reincarnate a1. Also, sync a1's membership with it.
+                dsl.changeStatus(t, tc, 0, 0, 'alive', 3),
+                dsl.waitForPingResponse(t, tc, 0),
+
+                // increase internal incarnation numbers of b1 and b2.
+                increaseInternalIncB(t, tc),
+
+                // Decrease the incarnation number of sut in B to 1000.
+                decreaseIncOfSutInB(t, tc),
+
+                // Clean up old pings before forcing the merge..
+                dsl.consumePings(t, tc),
+
+                // Now we have a clean partition with both sides reincarnated. According to sut:
+                // ● a1: 1340 alive.
+                // ● sut: current alive.
+                // ● b1: 1337 faulty.
+                // ● b2: 1337 faulty.
+                //
+                // Also, ∀ b ∈ B ⇔ {b1, b2}:
+                // ● a1: 1337 faulty.
+                // ● sut: 1000 faulty.
+                // ● b1: 1340 alive.
+                // ● b2: 1340 alive.
+
+                // Force SUT to merge the partition.
+                dsl.callEndpoint(t, tc, "/admin/healpartition/disco", {},
+                        function(eventBody) { validateHealRequest(t, tc, eventBody) }),
+                removeHealPartitionDiscoResponse(t, tc),
+
+                // SUT sends /join to b.
+                dsl.waitForJoins(t, tc, 1),
+
+                // Assert ping was sent to b with a1 and sut alive.
+                checkSutMergedWithB(t, tc),
+
+                // Make node membership global again, since partition is over.
+                function(list, cb) {
+                    delete tc.test_state['a1']['membership'];
+                    delete tc.test_state['b1']['membership'];
+                    delete tc.test_state['b2']['membership'];
+                    cb(list);
+                },
+                dsl.assertStats(t, tc, 4, 0, 0),
+            ]
+        })
+     );
+
+
+function waitForPingResponseFromTarget(t, tc) {
+    return function waitForPingResponseFromTarget(list, cb) {
+        return dsl.waitForPingResponse(t, tc, tc.test_state['target_idx'])(list, cb);
+    }
+}
+
+// Mark b1 and b2 faulty in a1.
+function markBFaultyInA(t, tc) {
+    return function markBFaultyInA(list, cb) {
+        var b1_port = tc.test_state['b1'].port,
```

Comment:
At this point I'm a bit confused, when is tc.test_state['b1'] used for tracking membership state and when is tc.fakeNodes[ix_for_b1].membership